IOMC/RandomEngine/src/XorShift128PlusAdaptor.cc

```diff
@@ -0,0 +1,213 @@
+#include "IOMC/RandomEngine/src/XorShift128PlusAdaptor.h"
+#include "FWCore/Utilities/interface/EDMException.h"
+
+#include "CLHEP/Random/engineIDulong.h"
+
+#include "FWCore/MessageLogger/interface/MessageLogger.h"
+
+#include "TBufferFile.h"
+
+#include <cstddef>
+#include <iomanip>
+#include <iostream>
+#include <sstream>
+#include <stdint.h>
+#include <string>
+#include <time.h>
+
+
+namespace edm {
+
+XorShift128PlusAdaptor::XorShift128PlusAdaptor() {
+  
+  srand(time(NULL));
+  for( auto& seed : seeds ) {
+    seed = rand();
+  }
+}
+
+XorShift128PlusAdaptor::XorShift128PlusAdaptor( const std::array<uint64_t,2>& inseeds ) {  
+  if( inseeds.size() < 2 ) {
+    Grumble(std::string("XorShift128Plus needs two 64 bit seeds!"));
+  }
+  
+  if( inseeds[0] == 0 && inseeds[1] == 0 ) {
+    Grumble(std::string("XorShift128Plus cannot be seeded with all zeroes!"));
+  }
+  seeds = inseeds;
+  state = seeds;
+  theSeeds = reinterpret_cast<const int64_t*>(seeds.data());
+}
+
+XorShift128PlusAdaptor::XorShift128PlusAdaptor( const std::array<uint32_t,4>& inseeds ) {  
+  if( inseeds.size() < 4 ) {
+    Grumble(std::string("XorShift128Plus needs four 32 bit seeds!"));
+  }
+  
+  if( inseeds[0] == 0 && inseeds[1] == 0 && 
+      inseeds[2] == 0 && inseeds[2] == 0    ) {
+    Grumble(std::string("XorShift128Plus cannot be seeded with all zeroes!"));
+  }
+
+  seeds.fill(0ULL);
+  for( size_t i = 0; i < inseeds.size() && i < 4; ++i ) {
+    uint64_t temp = inseeds[i];
+    seeds[i/2] = ( temp <<( i%2 ? 0 : 32 ) ); 
+  }
+  state = seeds;
+  theSeeds = reinterpret_cast<const int64_t*>(seeds.data());
+}
+
+XorShift128PlusAdaptor::XorShift128PlusAdaptor(std::istream&) {
+  Grumble(std::string("Cannot instantiate a TRandom engine from an istream"));
+}
```

Comment:
Both are factors slower than the xorshift. I'll post benchmarks.

Life gets difficult with 6.5 million cells.

-L

(Sent from my Nexus 6)
On Apr 6, 2016 8:34 PM, "W. David Dagenhart" <notifications@github.com>
wrote:

> In IOMC/RandomEngine/src/XorShift128PlusAdaptor.cc
> <https://github.com/cms-sw/cmssw/pull/13921#discussion_r58758069>:
>
> > +      inseeds[2] == 0 && inseeds[2] == 0    ) {
> > +    Grumble(std::string("XorShift128Plus cannot be seeded with all zeroes!"));
> > +  }
> > +
> > +  seeds.fill(0ULL);
> > +  for( size_t i = 0; i < inseeds.size() && i < 4; ++i ) {
> > +    uint64_t temp = inseeds[i];
> > +    seeds[i/2] = ( temp <<( i%2 ? 0 : 32 ) );
> > +  }
> > +  state = seeds;
> > +  theSeeds = reinterpret_cast<const int64_t*>(seeds.data());
> > +}
> > +
> > +XorShift128PlusAdaptor::XorShift128PlusAdaptor(std::istream&) {
> > +  Grumble(std::string("Cannot instantiate a TRandom engine from an istream"));
> > +}
>
> Should change the references from TRandom to XorShift128Plus. Just
> wondering, did you try TRandom3 instead of HepJames and was it also too
> slow? I do not know the answer, I was just wondering.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
> <https://github.com/cms-sw/cmssw/pull/13921/files/d50f69739b893fbcc30681627e2cca91bb593879#r58758069>
>